proposals/0003-external-calls.md

```diff
@@ -0,0 +1,344 @@
+Whether using raw calls (of the form `someAddress.call()`) or contract calls (of the form `ExternalContract.someMethod()`), malicious code might execute. Even if ExternalContract is not malicious, malicious code can be executed by any contracts it calls.
+One particular danger is malicious code may hijack the control flow, leading to race conditions.
+
+##### Re-entrancy
+- `someAddress.send()` and `someAddress.transfer()` in Solidity are considered safe against reentrancy. While these methods still trigger code execution, the called contract is only given a stipend of 2300 gas which is currently only enough to log an event.
+  - Prevents reentrancy but is incompatible with any contract whose fallback function requires 2300 gas or more.
+  - Sometimes the programmer won't want this, but then has to fall back onto the dangerous raw calls.
+
+#### 3. External calls can silently fail
+Solidity offers low-level call methods that work on rawAddress: `address.call()`, `address.callcode()`, `address.delegatecall()`, `address.send()`. These low-level methods never throw an exception so they fail silently.
+
+```javascript
+// bad
+someAddress.send(55);
+// this is doubly dangerous, as it will forward all remaining gas and doesn't check for result
+someAddress.call.value(55)(); 
+// if deposit throws an exception, the raw call() will only return false and transaction will NOT be reverted
+someAddress.call.value(100)(bytes4(sha3("deposit()"))); 
+
+// good
+if(!someAddress.send(55)) {
+    // Some failure code
+}
+
+ExternalContract(someAddress).deposit.value(100);
+```
+
+#### 4. Interfaces can easily be incorrectly specified
+Minor errors in Solidity interfaces may lead to wrong code being executed. `Alice.set(uint)` takes an `uint` in `Bob.sol` but `Alice.set(int)` takes an `int` in `Alice.sol`. The two interfaces will produce two differents method IDs. As a result, Bob will call the fallback function of Alice rather than of `set`. This type of error is responsible for the King of the Ether bug.
+
+- [King of the Ether](https://www.kingoftheether.com/postmortem.html) (line numbers:
+	[100](KotET_source_code/KingOfTheEtherThrone.sol#L100),
+	[107](KotET_source_code/KingOfTheEtherThrone.sol#L107),
+	[120](KotET_source_code/KingOfTheEtherThrone.sol#L120),
+	[161](KotET_source_code/KingOfTheEtherThrone.sol#L161))
+
+## Proposed solution
+The following solution is partially based upon the [Command Design Pattern]() and the [Oraclize Engine](https://docs.oraclize.it/). They allow for execution of the argument if other given conditions are met (as specified by the compiler). A valid external call should specify the following, some of these can be auto-filled by the compiler:
+- The contract address
+- The function name
+- The parameters
+- The gas allocation
+- The ether allocation
+
+Our motivations are:
+1. Contracts are untrustworthy by default;
+1. Arbitrary code may get executed;
+1. Failures may be silent;
+1. Interfaces may be incorrectly specified.
```

Comment:
These numbers should increment properly so it's nice to read in unrendered form.